### PR TITLE
follow-up #2261 (gpu fallbacks)

### PR DESCRIPTION
**Pull request by thundergolfer**
## Describe your changes

Follow-up to https://github.com/modal-labs/modal-client/pull/2261

**Demo**

text version of Irfan's example

```python
import subprocess

import modal

app = modal.App()


@app.function(gpu=["t4", "a10g"], max_inputs=1)
def fallback():
    subprocess.run("nvidia-smi --list-gpus", shell=True, check=True)


@app.local_entrypoint()
def main():
    for i in range(30):
        fallback.remote()
```

<details> <summary>Backward/forward compatibility checks</summary>

---

Check these boxes or delete any item (or this section) if not relevant for this PR.

- [x] Client+Server: this change is compatible with old servers
- [x] Client forward compatibility: this change ensures client can accept data intended for later versions of itself

Note on protobuf: protobuf message changes in one place may have impact to
multiple entities (client, server, worker, database). See points above.

---

</details>

## Changelog

* `modal.Function` and `modal.Cls` now support specifying a `list` of GPU configurations, allowing the Function's container pool to scale across each GPU configuration in preference order.

### CODE REVIEW

File: modal/app.py [hunk 1]
```diff
@@ -616,7 +616,7 @@ def function(
         schedule: Optional[Schedule] = None,  # An optional Modal Schedule for the function
         secrets: Sequence[_Secret] = (),  # Optional Modal Secret objects with environment variables for the container
         gpu: Union[
-            GPU_T, Sequence[GPU_T]
+            GPU_T, List[GPU_T]
```
Comment (thundergolfer) [inline modal/app.py:619]: I think this is better as we don't mean for the `Sequence` to be a tuple or a string or anything else that implements the sequence interface. We just want a plain `list`.

**Comment by thundergolfer** [comment]
@irfansharif I think we should include the change log message as soon as we merge the functionality. It's out on PyPi after all 🙂

File: modal/sandbox.py [hunk 2]
```diff
@@ -85,6 +85,12 @@ def _new(
                 raise InvalidError("`region` and `_experimental_scheduler_placement` cannot be used together")
             scheduler_placement = SchedulerPlacement(region=region)
 
+        if isinstance(gpu, list):
```
Comment (irfansharif) [inline modal/sandbox.py:88]: The typing for Sandbox._new doesn't take in the list form, but makes sense to surface a helpful error here anyway given the misuse possibility.

File: modal/app.py (hunk 1 above)
Comment (irfansharif) [reply modal/app.py:619]: Maybe `region: Optional[Union[str, Sequence[str]]] = None,` should be made as a `List` too - was doing it to stay internally consistent.

**Review by irfansharif — APPROVED** [review]

**Comment by irfansharif** [comment]
>  I think we should include the change log message as soon as we merge the functionality. It's out on PyPi after all 🙂

I wanted to wait until we had synmons running internally (https://github.com/modal-labs/modal/pull/15907) - thanks for following up here!

**Comment by mwaskom** [comment]
You can also manually update the `CHANGELOG.md` to attribute the feature to the correct version.

**Comment by thundergolfer** [comment]
> should be made as a List too - was doing it to stay internally consistent.

Yep I do think so, but wanted to keep that out of this diff. I also spotted an unrelated issue in `run_function` while reading the code.

**Pull request merged**